### ISSUE TITLE
Client mocking or intercepting?

### ISSUE THREAD

**Issue by SergeyTsaplin** [hyperium/hyper#1618]
It's really hard to write unit tests for the client usage, the library should provide public mocks for testing purposes

**Comment by madmaxio** [comment]
What do you mean by public mocks?

**Comment by rolftimmermans** [comment]
I imagine something like this: https://github.com/Byron/yup-hyper-mock

It's nice that this project exists, but it would be even nicer if it were included with hyper so that there will never be any incompatibilities.

**Comment by nanreh** [comment]
It would be nice to offer more test-ability than provided by `yup-hyper-mock`. Specifically, it seems like that crate mocks the `Connector` and that limits what the mock has access to. E.g. what if we want the mock to reply differently to `POST /something` and `GET /something`? Or based on a request header?

Ideally, we'd have a way to provide a custom `Service` when building a `Client` to match how we can do that when we build a `Server`. I don't see an obvious way to do that currently with Hyper.

There a bit more context in [this Rust lang forum question](https://users.rust-lang.org/t/how-can-i-intercept-hyper-client-requests-to-log-them/47649).

I'm pretty new to Rust so I apologize in advance if I'm missing an existing method for achieving this.

**Comment by sadokmtir** [comment]
Is there <ny updates regarding this issue ? Or did anyone managed to do it in a neat way to mock client calls ?

**Comment by seanmonstar** [comment]
There's nothing new here. I'm not sure there's anything for hyper to do.

In any application I was writing, if I needed to mock anything, I'd likely do that through a test framework. Like, if I needed to mock a `TcpStream`, I'd probably make the function I want to test be generic over `T: AsyncRead + AsyncWrite`, and then I could provide a `DuplexStream`, or something from `tokio-test`, or my own custom thing. I wouldn't expect `TcpStream` to have a way to mock out its internals.

So, I'm going to close this. But if there's ever an idea to have some sort of `hyper-test` crate with a proposal of what that looks like, we can open an issue/RFC.
